boa_engine/src/builtins/promise/mod.rs

```diff
@@ -0,0 +1,731 @@
+//! This module implements the global `Promise` object.
+
+#![allow(dead_code, unused_results, unused_variables)]
+
+#[cfg(test)]
+mod tests;
+
+mod promise_job;
+
+use boa_gc::{Finalize, Gc, Trace};
+use boa_profiler::Profiler;
+use tap::{Conv, Pipe};
+
+use crate::{
+    builtins::BuiltIn,
+    context::intrinsics::StandardConstructors,
+    job::JobCallback,
+    object::{
+        internal_methods::get_prototype_from_constructor, ConstructorBuilder, FunctionBuilder,
+        JsObject, ObjectData,
+    },
+    property::Attribute,
+    value::JsValue,
+    Context, JsResult,
+};
+
+use self::promise_job::PromiseJob;
+
+use super::JsArgs;
+
+/// JavaScript `Array` built-in implementation.
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct Array;
+
+#[derive(Debug, Clone, Trace, Finalize)]
+enum PromiseState {
+    Pending,
+    Fulfilled,
+    Rejected,
+}
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub struct Promise {
+    promise_result: Option<JsValue>,
+    promise_state: PromiseState,
+    promise_fulfill_reactions: Vec<ReactionRecord>,
+    promise_reject_reactions: Vec<ReactionRecord>,
+    promise_is_handled: bool,
+}
+
+#[derive(Debug, Clone, Trace, Finalize)]
+pub struct ReactionRecord {
+    promise_capability: Option<PromiseCapability>,
+    reaction_type: ReactionType,
+    handler: Option<JobCallback>,
+}
+
+#[derive(Debug, Clone, Trace, Finalize)]
+enum ReactionType {
+    Fulfill,
+    Reject,
+}
+
+#[derive(Debug, Clone, Trace, Finalize)]
+struct PromiseCapability {
+    promise: JsValue,
+    resolve: JsValue,
+    reject: JsValue,
+}
+
+#[derive(Debug, Trace, Finalize)]
+struct PromiseCapabilityCaptures {
+    promise_capability: Gc<boa_gc::Cell<PromiseCapability>>,
+}
```

Comment:
I would like to have some documentation for these structures, to understand their purpose, if there is any reference to them in the spec, and `ReactionType` and `PromiseState` can probably implement `Copy`, which would make things easier.

For `Copy` types, you will probably need to implement `Trace` and `Finalize` separately, with an empty implementation, or better: whenever a structure requires them in, use the `#[unsafe_ignore_trace]` attribute on the field.